docs: VS code debugging config & MRE

MentalGear opened the pull request:
Vs Code debug config

MentalGear commented:
As proposed by @erezrokah
https://github.com/MentalGear/netlify-dev_debug_vs-code/pull/1#issuecomment-846510622

erezrokah commented:
Sorry for the force push @MentalGear, I though the missing status checks are because you were missing a rebase, but it's because of https://github.com/netlify/cli/blob/a2e10672b4695cd69d430e0a263d7c3e1d0175b5/.github/workflows/main.yml#L11

MentalGear commented:
> Sorry for the force push @MentalGear, I though the missing status checks are because you were missing a rebase, but it's because of
> 
> https://github.com/netlify/cli/blob/a2e10672b4695cd69d430e0a263d7c3e1d0175b5/.github/workflows/main.yml#L11

NP. So this pull request should be good to go, right ?

erezrokah commented:
Sorry for the delay @MentalGear, I'll review it today

erezrokah commented on docs/functions-dev.md:
````diff
@@ -42,3 +42,66 @@ To do so set the `--inspect` Node.js option when starting the functions server:
+
+```jsonc
+{
+  // Use IntelliSense to learn about possible attributes.
````
Can we remove the code comments to brevity? 
It seems we don't need the `https://go.microsoft.com/fwlink/?linkid=830387` and the `FULLSTACK NETLIFY DEV DEBUGGING` steps are already described above

erezrokah replied:
I think we can remove both in favor of a `to learn more about VSCode launch configurations, see (here)[https://code.visualstudio.com/docs/editor/debugging#_launch-configurations]`

Or something similar. That way we can be brief and provide users the additional information without duplicating it.

MentalGear replied:
All right, I will add link towards the official vs code guide and the netlify guide.

erezrokah commented on docs/functions-dev.md:
```diff
@@ -42,3 +42,66 @@ To do so set the `--inspect` Node.js option when starting the functions server:
+  // 2. Run 'Fullstack Debugging' from the debug interface
+  // 3. A chrome browser window will open, however it may take a few seconds
+  // before the server is ready and content is displayed.
+  "compounds": [
```
Do we need the Chrome client? This ties the configuration to a specific browser which might not be installed on some machines. Also I was prompted to install a VS Code extension to get it working.

I was able to debug the CLI by using the following minimal config:
```json
{
  "version": "0.2.0",
  "configurations": [
    {
      "type": "node",
      "request": "launch",
      "name": "Launch Program",
      "skipFiles": ["<node_internals>/**"],
      "program": "${workspaceFolder}/node_modules/.bin/netlify",
      "args": ["dev"]
    }
  ]
}
```

Also we should instruct users to install the CLI locally via `npm install --save-dev netlify-cli` for this to work

erezrokah replied:
This section is about functions debugging and I see it serving 2 purposes:
1. Debugging only functions with `netlify functions:serve`
2. Debugging functions with redirects, headers, etc with `netlify dev`

As for frontend debugging, I don't think it's related to the original issue https://github.com/netlify/cli/issues/409, so can we descope it from this PR?

I agree this is a valid use case, so having another section for it would makes sense.
In that section I prefer to link to https://code.visualstudio.com/blogs/2016/02/23/introducing-chrome-debugger-for-vs-code or https://code.visualstudio.com/docs/editor/debugging instead of duplicating content that might go stale.

Please let me know if that makes sense, and sorry for not being clear enough on this from the start.

MentalGear replied:
Thanks for clarifying that as indeed I overlooked that the page is called "Serving functions locally" with only the subsection being about local function debugging.

Since you also agree that having fullstack debugging for netlify-cli is valid, I would then rather propose to simply spawn out the whole debugging topic into it's own page where we can have both functions (backend) and frontend (chrome client) debugging mentioned. What do you think ?

erezrokah replied:
> Since you also agree that having fullstack debugging for netlify-cli is valid, I would then rather propose to simply spawn out the whole debugging topic into it's own page where we can have both functions (backend) and frontend (chrome client) debugging mentioned. What do you think

Yes that makes sense.

erezrokah commented on docs/functions-dev.md:
```diff
@@ -42,3 +42,66 @@ To do so set the `--inspect` Node.js option when starting the functions server:
+      "request": "launch",
+      "skipFiles": ["<node_internals>/**"],
+      "program": "${workspaceFolder}/node_modules/.bin/netlify",
+      "args": ["dev", "--inspect"],
```
I believe `"--inspect"` can be removed

MentalGear replied:
You're the expert on the internal working of it, so if it serves no (upcoming) purpose I'll remove it.

erezrokah replied:
The CLI ignores that flag so it serves no purpose (it's not a CLI flag, it's a Node.js flag).

erezrokah commented on docs/functions-dev.md:
```diff
@@ -42,3 +42,66 @@ To do so set the `--inspect` Node.js option when starting the functions server:
+      "type": "node",
+      "request": "launch",
+      "skipFiles": ["<node_internals>/**"],
+      "program": "${workspaceFolder}/node_modules/.bin/netlify",
```
As mentioned before this only works after installing the CLI locally so we should instruct users to do so.

erezrokah commented on docs/functions-dev.md:
```diff
@@ -42,3 +42,66 @@ To do so set the `--inspect` Node.js option when starting the functions server:
+      "skipFiles": ["<node_internals>/**"],
+      "program": "${workspaceFolder}/node_modules/.bin/netlify",
+      "args": ["dev", "--inspect"],
+      "resolveSourceMapLocations": [
```
I don't think this is required as a default configuration.

I would rather have another section that explains about source maps and the purpose of this setting, as this is a more advanced use case.

MentalGear replied:
As I understand, netlify dev is used with js/ts/node. I would think then that a) using node modules would be pretty default for this use case b) as is the need to exclude them from source maps.

For me at least, without this source-map config, I get lots of errors in the debug console on startup.  I remember also reading that without proper sourcemap config breakpoints might not work (I didn't test that).

However, also happy to spawn this out into it's own section if you prefer. :)

What concern me a more however is, that I noticed that errors are not forwarded to `debug console` or vs code.

erezrokah replied:
> As I understand, netlify dev is used with js/ts/node

I'm not sure I understand this statement. `netlify dev` runs as a regular Node.js executable and doesn't require any source maps.

> For me at least, without this source-map config, I get lots of errors in the debug console on startup. I remember also reading that without proper sourcemap config breakpoints might not work (I didn't test that).
> 
> However, also happy to spawn this out into it's own section if you prefer. :)

Could this be related to frontend debugging? Have you seen this for functions too? If so, can you share an example repo? By default, breakpoints in JavaScript functions should work.

MentalGear replied:
I think VS code debugger may need access to the sourcemaps.

The error messages are showing up in the netlify dev debug console, not the client console, so definitely backend related.

You can reproduce it by using my MRE https://github.com/MentalGear/netlify-dev_debug_vs-code/issues and comment out in `launch.json`:
- `resolveSourceMapLocations `
- "console": "integratedTerminal" (this I added later as vs-code `debug console` doesn't forward all the std output as integratedTerminal does)

```jsonc
      // "console": "integratedTerminal" // avoids logs & errors from being suppressed
      // "resolveSourceMapLocations": [
      //   // NOTE: your linter may think this property shouldn't be here
      //   // yet it works for finding the source maps
      //   "${workspaceFolder}/**",
      //   "!**/node_modules/**"
      // ]
```

erezrokah reviewed (changes requested):
Hi @MentalGear, thank you for submitting the PR and sorry for the late reply.

I added some comments targeted at simplifying the default configuration as much as possible.

Please let me know what you think.

MentalGear commented:
Thanks for the review. I added to your comments above and updated the guide with the dependency note as well as a more straightforward numbered instruction.

erezrokah commented on docs/functions-dev.md:
```diff
@@ -42,3 +42,80 @@ To do so set the `--inspect` Node.js option when starting the functions server:
+## VS Code Debugging
+
+### 1. Setup
+Install netlify dev as a local dev dependency:
```
```suggestion
Install Netlify CLI as a local dev dependency:
```

erezrokah reviewed (changes requested):
Thanks for the follow up @MentalGear.

I've added a few comments to clarify my intention with the initial review.

MentalGear reviewed (commented):
Thanks for replying.

I adjusted the content were necessary and replied to your questions. Overall, I'm in favor though to spawn out the whole debugging into its own page for best DX coverage. Let me know your thought.

MentalGear commented:
I have spawn out the Debugging part into its own page and added the discussed changes. @erezrokah.

erezrokah commented:
Sorry for the very late follow up @MentalGear. I was finally able to dig more into this.
I think I was able to further reduce the launch configurations size and also avoid the usage of an extension.
Also, now VSCode will wait for the server to be up before launching the browser.

I didn't have any issues with source maps not loading with my Create React App project, so I'd assume their project specific and would want to avoid adding them in the default configuration.

If that's not the case can you please  guide how to reproduce that issue?

I've pushed my changes in https://github.com/netlify/cli/pull/2440/commits/adfe78f4da253acd77a7ba2aa6e72006fe32fbfa

See https://deploy-preview-2440--cli.netlify.app/vscode

rstavchansky commented on docs/vscode.md:
```diff
@@ -0,0 +1,67 @@
+---
+title: Usage with Visual Studio Code
```
This may need a more specific title. Would this work?

Another option would be `Run and debug with Visual Studio Code`, per the comment below

```suggestion
title: Debugging with Visual Studio Code
```

rstavchansky commented on docs/vscode.md:
```diff
@@ -0,0 +1,67 @@
+title: Usage with Visual Studio Code
+---
+
+# Usage with Visual Studio Code
```
```suggestion
# Run and debug with Visual Studio Code
```

rstavchansky commented on docs/vscode.md:
```diff
@@ -0,0 +1,67 @@
+# Usage with Visual Studio Code
+
+This guide will walk you through configuring VScode to run and debug `netlify dev`.
+That allows debugging your site and functions all in the same IDE.
```
Small suggested wording tweak: 

```suggestion
This allows debugging your site and serverless functions all in the same IDE.
```

rstavchansky commented on site/src/_app.js:
```diff
@@ -67,6 +68,7 @@ const pageNames = {
   contributing: 'Contributing',
   'netlify-dev': 'Netlify Dev',
   'functions-dev': 'Functions Development',
+  vscode: 'Usage with VSCode',
```
If changed above - 

```suggestion
  vscode: 'Run and debug with VSCode',
```

rstavchansky reviewed (commented):
Adding some minor suggestions from a docs perspective
